### DISCUSSION TITLE
aerial imagery tiles colour correction

### DISCUSSION THREAD

**Discussion by RobShaw77** [ImageMagick/ImageMagick#5296] [Help] [answered]
I have a set of aerial imagery tiles (about 130,000) for England (infrared imagery, although that doesn't matter directly). The problem I have is that in some cases adjacent tiles have completely different colour hue as they were taken on different dates or at different times of the day.  Is it possible to use ImageMagick to 'normalise' the colour between adjacent tiles so that the transition between them is smooth?

![SE3330](https://user-images.githubusercontent.com/29256807/177112158-1bd01f18-7a7a-45e6-8099-bae425661a95.jpg)
![SE3329](https://user-images.githubusercontent.com/29256807/177112147-8160f0ae-df76-4541-9ee5-a29aab8decd5.jpg)

Thanks for any help.

Rob

**Comment by snibgo** [comment]
If your images had an overlap, I would have a method for that. The method is called "seamless montage" or "seamless blend". When there is no overlap, I don't know how to do it.

Let us pretend that your two 2000x2000 pixel images had an overlap of 10 pixels, so the final image will be 2000x3990 pixels, and that we want to adjust the second image to match the first. Windows command:

```
magick ^
  -background White ^
  ( aerial1.jpg -gravity North -extent 2000x3990 ) ^
  ( aerial2.jpg -gravity South -extent 2000x3990 ) ^
  ( -size 2000x2000 xc:Black -size 2000x1990 xc:White -append +repage ) ^
  -define compose:args=400x0.00002+100 ^
  -geometry +0+10 ^
  -compose seamless-blend ^
  -verbose ^
  -composite ^
  +verbose ^
  out.jpg
```
out.jpg (reduced in bytes for the web) is:

![x](https://user-images.githubusercontent.com/33812368/177173361-eabcc2d7-24ae-485d-9963-e1581891aca5.jpg)
The result is wrong because we don't really have an overlap. But it isn't badly wrong.

On "-compose seamless-blend" see https://github.com/ImageMagick/ImageMagick/discussions/4712

For a lower-level description of the method, see my page [Seamless photomontage](http://im.snibgo.com/seamlpm.htm).

> I have a set of aerial imagery tiles (about 130,000) for England ...

I am curious about this (I live in England). Is that data public?

**Reply by RobShaw77** [reply]
Many thanks for your reply. Is there a way with Magick to buffer out (ie repeat) the edge pixels of each image? That might give the overlap you describe.

Re the data: it's from Blue Sky and accessible as part of the Public Sector Geospatial Agreement (I work for NHS England). I can share a small local area with you for test purposes if there is a particular place you are interested in.

**Comment by snibgo** [comment]
Here is a better result. It duplicates the top row of the second image, by cropping that row and appending it to the top. It assumes the bottom row of the first image should match the new top row of the second image.

So the result is spatially aligned, and the bottom half has colours adjusted to roughly match the top half.

Windows BAT script:
```
magick ^
  -background White ^
  ( aerial1.jpg ^
    -gravity North -extent 2000x4000 ^
  ) ^
  ( aerial2.jpg ^
    ( +clone -gravity North -crop x1+0+0 +repage ) ^
    +swap ^
    -append +repage ^
    -gravity South -extent 2000x4000 ^
  ) ^
  ( -size 2000x2000 xc:Black -size 2000x2000 xc:White -append +repage ) ^
  -define compose:args=400x0.00002+100 ^
  -geometry +0+0 ^
  -compose seamless-blend ^
  -verbose ^
  -composite ^
  +verbose ^
  out2.jpg
```

![out2](https://user-images.githubusercontent.com/33812368/177355144-d0e17792-faca-4320-ab47-f845dd9adb08.jpg)

We can modify "-crop x1+0+0" to "-crop x10+0+0" to use ten rows for matching, or whatever number we want. I have tried 1, 10 and 100, and I think the result from 1 is best.

Matching can't be perfect because the sun is in a different position, so different parts of buildings receive different relative illumination, etc.

In theory, whoever made this data (Bluesky?) should be able to supply images with overlap. This would enable more accurate matching. But they might not want to for commercial reasons.

Data seems to be available in TIFF format, rather than JPG. That may also increase accuracy, but the difference may not be visible, and the files would be much larger.

(The NHS has saved my life more than once. I'm a big fan.)

**Comment by snibgo** [comment]
Alternative methods ignore the fact that the images should abut each other, and simply tweak colours to make one look like the other.

One method for this is by matching histograms.
```
call %PICTBAT%matchHisto aerial2.jpg aerial1.jpg a3.tiff

%IMG7%magick ^
  aerial1.jpg a3.tiff ^
  -append +repage ^
  out3.jpg
```
See [Process modules: matching histograms](http://im.snibgo.com/customim.htm#matchhist). This needs IM to be built with my process modules.

See out3.jpg below.

A simpler version merely matches the mean and standard deviation for each channel.

```
call %PICTBAT%imgGainBias aerial2.jpg aerial1.jpg a4.tiff gain

%IMG7%magick ^
  aerial1.jpg a4.tiff ^
  -append +repage ^
  out4.jpg
```
See out4.jpg below.

Even simpler, just match the mean:
```
call %PICTBAT%setMeanImg aerial2.jpg aerial1.jpg a5.tiff

%IMG7%magick ^
  aerial1.jpg a5.tiff ^
  -append +repage ^
  out5.jpg
```
See out5.jpg below.

My Windows scripts are available at im.snibgo.com

My overall suggestion: if overlaps are available, "-compose seamless-blend" should work well. But if you don't have overlaps, matching histograms works okay.


out3.jpg is:
![out3](https://user-images.githubusercontent.com/33812368/177391135-8721fa10-bfc3-4621-8052-79471a08b6d4.jpg)

out4.jpg is:
![out4](https://user-images.githubusercontent.com/33812368/177391163-d40b7c0e-77ba-440e-9220-e5d5fb789f4a.jpg)

out5.jpg is:
![out5](https://user-images.githubusercontent.com/33812368/177391197-5f958fb9-4a1a-4eb8-8eda-b5701f30eccc.jpg)

**Reply by RobShaw77** [reply]
Thank you this is great. The histogram matching looks like it might be the best approach given how many images I have to process.

Unfortunately I've failed completely in building IM (./configure etc) with the modules option using cygwin. It appears to run fine but hasn't created any of the files :(

**Comment by snibgo** [comment, marked as answer]
A strategy for dealing with 130,000 images depends on your objectives. This might include: automated analysis to determine amount of foliage, buildings, roads etc. Or to normalise the complete set so any particular subset can be appended and will "look good". Or to have a process for extracting any particular subset, and processing just those images so when they are appended they "look good". And so on.

On matching histograms (or matching mean and SD, or matching just the mean): this assumes that the images should look roughly like each other. This is true in the two examples you show. But we might have one image that is mostly foliage (so it is mostly red) and another image with no foliage (so no red). In this case, matching histograms (or anything else) won't work well.

To deal with this, and enable simple automated processing of 130,000 images, the following may work:

1. Calculate the ideal histogram (or ideal mean and SD, etc) for images that are entirely foliage, and images that have no foliage.

2. For each image in the set, analyse it to calculate the proportion that is foliage. Then the ideal histogram (or mean, etc) is an interpolation between the histogram for 100% foliage and the histogram for zero foliage. Process the image to apply that ideal.

On building IM with my process modules: I assume you have a Windows computer, with Cygwin. Is that correct? And you are following instructions on [Process modules](http://im.snibgo.com/customim.htm)?

In a nutshell, the method is to download my filter source code files, including `Makefile.am` into the `filters` subdirectory, and run the bash commands `automake`, `autoconfig`, `./configure`, `make`, `make install`. During `make`, you should see each of my filters being compiled.

Feel free to post specific questions about the build process.
